### PR TITLE
BREAKING CHANGE: remove `node-fetch` in favor of global

### LINKED ISSUE #432: pin version of `node-fetch` and add end-to-end test to catch breaking changes
This is in response to https://github.com/node-fetch/node-fetch/issues/1300

### LINKED ISSUE #456: Use an isomorphic fetching library
I've recently been trying to use an edge environment to deploy a project of mine, and the main error that I get is: `Cannot bundle Node.js built-in "stream" imported from "node_modules/@octokit/request/node_modules/node-fetch/lib/index.mjs"`. However, edge environments do not support Node.js, even though I believed this was an isomorphic library. Apparently, `node-fetch` is not isomorphic. I don't know how hard this would be or how much time it would take, but I could also open a PR if necessary.

### PR DESCRIPTION
**Pull request by styfle**
- Resolves #456
- Resolves #422
- Resolves #432

----

### Before the change?

Previously you could run this on Node.js 14 and 16 without manually polyfilling `fetch`.

### After the change?

With this PR it will work in Node.js 18, browsers, and any JS runtime that provides `fetch`. If you want to run in Node.js 14 or 16, you can polyfill `globalThis.fetch` or pass in the request options.

### Other information

This improves compatibility with many runtimes because `node-fetch` is not the same as the standard `fetch` global.

----

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [x] Yes (Please add the `Type: Breaking change` label)

If `Yes`, what's the impact:

If you want to run in Node.js 14 or 16, you must polyfill `globalThis.fetch` or pass in the request options.

### Pull request type

Please add the corresponding label for change this PR introduces:
- Feature/model/API additions: `Type: Feature`

----

### CODE REVIEW

**Comment by wolfy1339** [comment]
Can you fix the tests for Node 14 and 16?

**Comment by wolfy1339** [comment]
I think adding a simple patch like so will work:
```js
(async function () {
  if (parseInt(process.versions.node.split(".")[0]) <= 18) {
    // @ts-expect-error
    globalThis.fetch = await import("node-fetch");
  }
})();
```

**Comment by wolfy1339** [comment]
The patch shouldn't go in any production code, just for the test files.

**Comment by styfle** [comment]
@wolfy1339 I used a similar solution but with `require()` so its synchronous a little less code.

I confirmed the tests are now passing on node 16 and 18.

**Comment by wolfy1339** [comment]
Even with `node-fetch-native` it wouldn't be a single change.

As `node-fetch` v3 drops the timeout option since it isn't in the spec, the types and code need updating.

**Comment by wolfy1339** [comment]
We are using `node-fetch` v2 at the moment which had the timeout option.
In v3 it was removed.

https://github.com/node-fetch/node-fetch/blob/main/docs/v3-UPGRADE-GUIDE.md#the-timeout-option-was-removed

**Comment by gr2m** [comment]
I would remove the dependency on `node-fetch` altogether. That's what we already did in octokit-next:
https://github.com/octokit/octokit-next.js/pull/62/files

A custom `fetch` function can be configured already. For users of older Node versions, we can document how to use `node-fetch` instead. 

Node 16 is the last version that doesn't have it built in, and it will be out of maintenance mode in September. After that all current JavaScript runtime environments have the global fetch method.

**Comment by wolfy1339** [comment]
We'll need to remove the `timeout` option as well, since it doesn't exist on the fetch API or newer versions of `node-fetch`.

**Comment by wolfy1339** [comment]
Since this is going to be a breaking change, let's just remove that option from our code altogether

**Comment by gr2m** [comment]
note that this will also be a breaking change for all the packages that depend on `@octokit/request`. If possible, I'd try to coordinate with other breaking changes in the dependents.

File: test/request.test.ts [hunk 1]
```diff
@@ -447,7 +451,7 @@ x//0u+zd/R/QRUzLOw4N72/Hu+UG6MNt5iDZFCtapRaKt6OvSBwy8w==
   });
 
   it("passes node-fetch options to fetch only", () => {
```
Comment (wolfy1339) [inline test/request.test.ts:453]: Remove `timeout` parameter tests

**Comment by wolfy1339** [comment]
Will this resolve #432 @gr2m as `node-fetch` won't be included anymore

**Comment by gr2m** [comment]
> Will this resolve #432 @gr2m as `node-fetch` won't be included anymore

yes 🚀

File: README.md [hunk 2]
```diff
@@ -10,8 +10,7 @@ to interact with [GitHub’s REST API](https://developer.github.com/v3/) and
 [GitHub’s GraphQL API](https://developer.github.com/v4/guides/forming-calls/#the-graphql-endpoint).
 
 It uses [`@octokit/endpoint`](https://github.com/octokit/endpoint.js) to parse
-the passed options and sends the request using [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)
-([node-fetch](https://github.com/bitinn/node-fetch) when the runtime has no native `fetch` API).
+the passed options and sends the request using [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).
```
Comment (gr2m) [inline README.md:13]: Shall we mention in the intro that a custom `fetch` function can be passed? Some thing like this

```suggestion
the passed options and sends the request using [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). You can pass a custom `fetch` function using the `request.fetch` option, see below.
```

File: src/fetch-wrapper.ts [hunk 3]
```diff
@@ -26,10 +25,10 @@ export default function fetchWrapper(
...
-  const fetch: typeof nodeFetch =
-    (requestOptions.request && requestOptions.request.fetch) ||
-    globalThis.fetch ||
-    /* istanbul ignore next */ nodeFetch;
+  let { fetch } = globalThis;
+  if (requestOptions.request?.fetch) {
+    fetch = requestOptions.request.fetch;
+  }
 
   return fetch(
```
Comment (gr2m) [inline src/fetch-wrapper.ts:33]: if neither `requestOptions.request.fetch` nor `globaltThis.fetch` is set, we should through a helpful error message.

File: test/request.test.ts [hunk 4]
```diff
@@ -1,9 +1,13 @@
+if (parseInt(process.versions.node.split(".")[0]) < 18) {
+  const fetch = require("node-fetch");
+  globalThis.fetch = fetch;
+  globalThis.Headers = fetch.Headers;
+}
```
Comment (gr2m) [inline test/request.test.ts:5]: I think we can drop this, as we will stop testing in Nod 16 and below as per https://github.com/octokit/request.js/pull/586

**Review by gr2m — COMMENTED** [review]
I would recommend we also remove the `request.timeout` option while at it. We already recommend `options.request.signal` today

File: src/fetch-wrapper.ts (hunk 3 above)
Comment (styfle) [reply src/fetch-wrapper.ts:37]: Added a new error in [93869aa](https://github.com/octokit/request.js/pull/580/commits/93869aae8bf019616a0385f691ccead11988e819)

File: test/request.test.ts (hunk 4 above)
Comment (styfle) [reply test/request.test.ts]: Removed in [7b2709e](https://github.com/octokit/request.js/pull/580/commits/7b2709e7fab54b470731e0985bc0158497821950)

File: test/request.test.ts (hunk 1 above)
Comment (styfle) [reply test/request.test.ts:462]: Removed it from the tests in [ea7c2ab](https://github.com/octokit/request.js/pull/580/commits/ea7c2abf255e056eca4aa35a8cb92a06802c794c)

**Comment by styfle** [comment]
Looks like I have some merge conflicts now. Do you prefer `git rebase` or `git merge` to resolve?

**Comment by wolfy1339** [comment]
Either way is fine. We squash and merge anyways in the end.

**Comment by gr2m** [comment]
Merge is easier to keep track of new changes for reviewers.

File: test/request.test.ts (hunk 1 above)
Comment (wolfy1339) [reply test/request.test.ts:462]: These tests don't really seem useful now. Could you remove them

Comment (styfle) [reply test/request.test.ts]: Sounds good! Removed in [3cfc7c8](https://github.com/octokit/request.js/pull/580/commits/3cfc7c87c935a82987ff5a3d0227f35998411b1e)

**Comment by styfle** [comment]
@wolfy1339 @gr2m I fixed the merge conflicts and addressed the remaining comments, thanks!

**Comment by wolfy1339** [comment]
There are some linting issues left to fix.

**Review by gr2m — APPROVED** [review]
Great work, just some lint errors left:
https://github.com/octokit/request.js/actions/runs/5039405053/jobs/9038110405?pr=580#step:5:7

running `npm run lint:fix` should do the trick

**Review by wolfy1339 — APPROVED** [review]

**Comment by gr2m** [comment]
Wow this is a big day for Octokit's cross plattform support. Thank you all for making this a reality ❤️

**Pull request merged**